### ISSUE TITLE
warn-unreachable false-positive when method updates attribute after assert

### ISSUE THREAD

**Issue by jwodder** [python/mypy#11969]
Consider the following code:

```python
from dataclasses import dataclass


@dataclass
class Foo:
    value: int
    modified: bool = False

    def update(self) -> None:
        self.value += 1
        self.modified = True


foo = Foo(42)
assert not foo.modified
foo.update()
assert foo.modified
print("Reached")
```

When this is run, the `print("Reached")` line is reached, but mypy doesn't seem to realize that.  Running `mypy --warn-unreachable --pretty` on the above code gives:

```
unreachable03.py:18: error: Statement is unreachable
    print("Reached")
    ^
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: both v0.931 and the GitHub version as of commit 48d810d5ed25f898d4f220a7476b6b267c289b9a
- Mypy command-line flags: `--warn-unreachable --pretty`
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: 3.9.9
- Operating system and version: macOS 11.6

**Comment by KotlinIsland** [comment]
In this scenario mypy pretty much assumes everything is immutable and only takes into consideration explicit mutation, not side effects from any function calls.

This is because otherwise most narrowing would be invalidated instantly.

And to check each function to see if and what variables it mutates would be extremely slow and complicated.

**Comment by jwodder** [comment]
Methods that mutate attributes are surely a common occurrence, and I feel that assuming that they don't exist for the purpose of detecting unreachable code is the wrong thing to do, as it would lead to too many false positives.  I don't know what sort of narrowing would be invalidated by not assuming everything is immutable, but could it be possible to only "invalidate" it for warn-unreachable purposes?

**Comment by jbcpollak** [comment]
I agree with @jwodder - this is common and its a bad assumption variables can't change. Typescript for example also has type narrowing and does not have this issue: https://www.typescriptlang.org/docs/handbook/2/narrowing.html

> TypeScript follows possible paths of execution that our programs can take to analyze the most specific possible type of a value at a given position. It looks at these special checks (called type guards) and assignments, and the process of refining types to more specific types than declared is called narrowing.

**Comment by erictraut** [comment]
TypeScript also assumes that variables are not changed "behind the back" of the type checker in another execution scope. Without this assumption, it wouldn't be feasible to apply type narrowing to any member access expressions.

```python
class Foo {
    bar: number | string = 0;

    setBarToString() {
        this.bar = '';
    }
}

function test() {
    const f = new Foo();
    f.bar = 1;

    const val1 = f.bar; // Type of "val1" is revealed as "number"

    f.setBarToString();

    const val2 = f.bar; // Type of "val2" is still revealed as "number"
}
```

**Comment by KotlinIsland** [comment]
Kotlin handles this correctly by only narrowing `val` properties, and leaving `var` properties un-narrowed. Kotlin also narrows on so called 'scope functions'(the last part of this example)
```kt
class Foo(
    val foo: Any,
    var bar: Any,
)

fun test() {
    val f = Foo(1, 1);

    if (f.foo is String) {
        val v2 = f.foo // Type of "v2" is revealed as "String"
    }
    if (f.bar is String) {
        val v2 = f.bar // Type of "v2" is still revealed as "Any"
    }
    (f.bar as? String)?.let {
        val v2 = it // Type of "v2" is now revealed as "String"
    }
}
```

This way Kotlin is able to maintain perfect correctness while still achieving useful narrowing.

**Comment by fohrloop** [comment]
I asked about this in [SO](https://stackoverflow.com/questions/78211119/how-to-tackle-statement-is-unreachable-unreachable-with-mypy-when-setting-at) where user sytech introduced a workaround using type guard functions using [TypeGuard](https://typing.readthedocs.io/en/latest/spec/narrowing.html#typeguard). They're essentially functions that must return a boolean, and if they return True the first positional argument (or first after self in methods) is guaranteed to have type `TYPE` if the function return type annotation is `TypeGuard[TYPE]`. 

I wrote two helper functions, `object_is` and `is_instance`. You then would use them with assert statements. For example:

```python
import sys
from typing import Type, TypeVar

if sys.version_info < (3, 10):
    from typing_extensions import TypeGuard
else:
    from typing import TypeGuard

from dataclasses import dataclass


@dataclass
class Foo:
    value: int
    modified: bool = False

    def update(self) -> None:
        self.value += 1
        self.modified = True


_T = TypeVar("_T")


def object_is(obj, value: _T) -> TypeGuard[_T]:
    return obj is value


def is_instance(val, klass: Type[_T]) -> TypeGuard[_T]:
    return isinstance(val, klass)


foo = Foo(42)
assert object_is(foo.modified, False)
foo.update()
assert object_is(foo.modified, True)
print("Reached")
```

This uses [typing-extensions](https://pypi.org/project/typing-extensions/) to make the solution to work on older versions, as TypeGuard was introduced in Python 3.10.

**Comment by Jacobfaib** [comment]
This problem continues to plague otherwise valid code. Take the following reproducer (which even uses `@property`, something commonly used to indicate "this variable access runs code"):
```python
class Foo:
    @property
    def bool_prop(self) -> bool:
        pass


f = Foo()
assert f.bool_prop
assert not f.bool_prop
x = 0
```
```sh
$ mypy mypy_bug.py
mypy_bug.py: note: At top level:
mypy_bug.py:19:1: error: Statement is unreachable  [unreachable]
    x = 0
    ^~~~~
Found 1 error in 1 file (checked 1 source file)
```

Mypy's decision to not address this on the grounds that type narrowing would be broken is strange to say the least.

Python type checking already relies entirely on promises by the programmer. If my code says `def foo() -> Bar`, then `mypy` must believe that is true. In fact, if it is not, I can even lie to `mypy`:
```python
def foo() -> Bar:
    return typing.cast(Bar, 12345)
```
And `mypy` is _required_ to believe me! To now have `mypy` turn around and say "You have typed your function one way, but I can't take your word for it" is counterintuitive. 

Also it seems that with this feature `mypy` is now attempting to prove the _value_ of certain variables. This is not (IMO) within the purview of a _type_ checker. In these cases the types are all correct. It is not up to mypy to also decide if the _values_ are correct, since it obviously cannot prove they are wrong.

**Comment by hauntsaninja** [comment]
mypy is very unlikely to make any changes here.

If you want an easy way to reset what mypy thinks it knows about the attributes of a variable, you can do `foo = foo` or one of several other workarounds.

**Comment by zed** [comment]
@hauntsaninja  inserting `foo.modified = foo.modified` before the assert foo.modified` doesn't help (same error: unreachable). `safe_assert`, `object_is`, `operator.is_` worked in my case.

**Comment by hauntsaninja** [comment]
I said foo = foo

**Comment by rnestler** [comment]
Interestingly enough changing from `assert not foo.modified` to `assert foo.modified == False` fixes it as well.

I don't think it's well behaved to assume that a method call on an object (`foo.update()` in this case) can't change its properties (`foo.modified` in this case).

Tested with mypy 1.15 and 1.16.
